xalan/src/test/java/org/apache/xalan/VersionTest.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the  "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.xalan;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+import java.io.ByteArrayInputStream;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class VersionTest {
+
+  @ParameterizedTest(name = "{0} -> {2}")
+  @MethodSource("testReadPropertiesArgs")
+  public void testReadProperties(String ignoredName, String properties, String version) {
+    try (MockedStatic<Version> versionMock = Mockito.mockStatic(Version.class, Mockito.CALLS_REAL_METHODS)) {
+      versionMock
+        .when(Version::getPropertiesStream)
+        .thenReturn(new ByteArrayInputStream(properties.getBytes()));
+      assertEquals(version, Version.readVersionNumber());
+    }
+  }
+
+  private static Stream<Arguments> testReadPropertiesArgs() {
+    return Stream.of(
+      Arguments.of("single line without line feed", "version=1.2.3", "1.2.3"),
+      Arguments.of("single line with line feed", "version=4.5.D6-SNAPSHOT\n", "4.5.D6-SNAPSHOT"),
+      Arguments.of("multiple lines with version number", "foo=bar\nversion=7.8.9\nbaz=zot\n", "7.8.9"),
+      Arguments.of("single line without version number", "verXion=7.8.9\n", "0.0.0"),
+      Arguments.of("multiple lines without version number", "foo=bar\nverXion=7.8.9\nbaz=zot\n", "0.0.0")
+    );
+  }
+
+  @Test
+  public void testCannotReadProperties() {
+    try (MockedStatic<Version> versionMock = Mockito.mockStatic(Version.class, Mockito.CALLS_REAL_METHODS)) {
+      versionMock
+        .when(Version::getPropertiesStream)
+        .thenThrow(NullPointerException.class);
+      assertEquals("0.0.0", Version.readVersionNumber());
```

Comment:
Do you expect NPE from `getPropertiesStream`? The method is a single line only (~`.getResourceAsStream`), and Javadoc says the method returns `null` in case the value is not found. Is there a real value in testing "what happens if getPropertiesStream throws NPE"?

Reply:
> I do not expect a NPE, but the javadoc says that NPE will be thrown if the input parameter is null

The input parameter is `"version.properties"` string literal, so NPE must not happen.
Frankly, testing NPE there is overkill.

At the same time, "IO" exception occurs **while accessing the stream** rather than at `getResourceAsStream`, so if you want to test how it behaves in case the stream misbehaves, then you'd better have a stream that throws IO exception from its `read` call.

Reply:
Just in case, here's an `InputStream` that throws:

```java
new InputStream() {
  @Override
  public int read() throws IOException {
    throw new IOException("not implemented");
  }
};
```

In other words, testing IO failure here with regular Java is not even harder than testing with Mockito.

Reply:
> The input parameter is `"version.properties"` string literal, so NPE must not happen. Frankly, testing NPE there is overkill.

Please read more carefully. I said the NPE in the test is just a stand-in for **any** runtime exception which could happen there. It was just an arbitrary choice, because it does not matterwhich exception I am throwing,as all exceptions are cought. So we are covering the catch block, which is the objective here. That should be obvious enough.

Reply:
> At the same time, "IO" exception occurs **while accessing the stream** rather than at `getResourceAsStream`, 

Again, read more carefully. I was talking about I/O exception due to using try with resources.

![image](https://github.com/apache/xalan-java/assets/1537384/4ea2e7a0-d281-4a80-856b-6492b382772f)

> so if you want to test how it behaves in case the stream misbehaves

I am testing anything that can go wrong in the whole try-block including the statement in parentheses right on top.

Reply:
> Just in case, here's an `InputStream` that throws

Thanks, irrelevant again.

> In other words, testing IO failure here with regular Java is not even harder than testing with Mockito.

Wrong. I/O happens in method Version::getPropertiesStream, and there I cannot inject anything unless I factor out that one line of code again, which is a recursive problem. At the end, I need to stub that method if I want to avoid real I/O to happen in the test, which is exactly my objective: I do not want the test to read any test resource file but run the test in memory. The by far simplest method is to simply stub that one-line method. Is that so hard to understand? It is almost too trivial to explain IMO.

Reply:
We're getting into a bad pattern of trying to design in reviews. (I've been guilty of it too.)

May I suggest that we keep review focused on "is it an improvement" and "is there any reason _not_ to commit", and move extended discussion back to the dev list?

That will let us merge sooner and with less noise. One can always submit another PR later if we agree there's an improvement.

Don't let the quest for perfect drive out what's good enough for now. Incremental steps.

Reply:
> We're getting into a bad pattern of trying to design in reviews. (I've been guilty of it too.)
> 
> May I suggest that we keep review focused on "is it an improvement" and "is there any reason _not_ to commit", and move extended discussion back to the dev list?

Yes, **pleeease!**

Joe, do you remember my post on the mailing list about dysfunctional, micro-managing code reviews in OSS projects? This is exactly what I was talking about.